src/sentry/profiles/flamegraph.py

```diff
@@ -27,10 +31,107 @@ def get_profiles_id(
+    return raw_snql_query(
         snql_query,
-        referrer=Referrer.API_PROFILING_PROFILE_FLAMEGRAPH.value,
+        referrer,
     )["data"]
+
+
+def get_profile_ids(
+    params: ParamsType,
+    query: Optional[str] = None,
+) -> Dict[str, List[str]]:
+    data = query_profiles_data(
+        params,
+        Referrer.API_PROFILING_PROFILE_FLAMEGRAPH.value,
+        selected_columns=["profile.id"],
+        query=query,
+    )
     return {"profile_ids": [row["profile.id"] for row in data]}
+
+
+def get_span_intervals(
+    project_id: str,
+    span_group: str,
+    transaction_ids: List[str],
+    organization_id: str,
+    params: ParamsType,
+) -> Dict[str, Any]:
+    query = Query(
+        match=Entity(EntityKey.Spans.value),
+        select=[
+            Column("transaction_id"),
+            Column("start_timestamp"),
+            Column("start_ms"),
+            Column("end_timestamp"),
+            Column("end_ms"),
+        ],
+        where=[
+            Condition(Column("project_id"), Op.EQ, project_id),
+            Condition(Column("transaction_id"), Op.IN, transaction_ids),
+            Condition(Column("group"), Op.EQ, span_group),
+            Condition(Column("timestamp"), Op.GTE, params["start"]),
+            Condition(Column("timestamp"), Op.LT, params["end"]),
+        ],
+    )
+    request = Request(
+        dataset=Dataset.SpansIndexed.value,
+        app_id="default",
+        query=query,
+        tenant_ids={
+            "referrer": Referrer.API_STARFISH_PROFILE_FLAMEGRAPH.value,
+            "organization_id": organization_id,
+        },
+    )
+    return raw_snql_query(
+        request,
+        referrer=Referrer.API_STARFISH_PROFILE_FLAMEGRAPH.value,
+    )["data"]
+
+
+def get_profile_ids_with_spans(
+    organization_id: str,
+    project_id: str,
+    params: ParamsType,
+    span_group: str,
+    query: Optional[str] = None,
+):
+    data = query_profiles_data(
+        params,
+        Referrer.API_STARFISH_PROFILE_FLAMEGRAPH.value,
+        selected_columns=["id", "profile.id"],
+        query=query,
+        additional_conditions=[
+            Condition(Function("has", [Column("spans.group"), span_group]), Op.EQ, 1)
+        ],
+    )
+    # map {transaction_id: (profile_id, [span intervals])}
+    transaction_to_prof = {row["id"]: (row["profile.id"], []) for row in data}
+
+    data = get_span_intervals(
+        project_id,
+        span_group,
+        list(transaction_to_prof.keys()),
+        organization_id,
+        params,
+    )
+
+    for row in data:
+        start_ns = (int(datetime.fromisoformat(row["start_timestamp"]).timestamp()) * 10**9) + (
+            row["start_ms"] * 10**6
+        )
+        end_ns = (int(datetime.fromisoformat(row["end_timestamp"]).timestamp()) * 10**9) + (
+            row["end_ms"] * 10**6
+        )
+        transaction_id = row["transaction_id"]
+
+        transaction_to_prof[transaction_id][1].append({"start": str(start_ns), "end": str(end_ns)})
```

Comment:
There's no guarantee that the transaction id you get from the span table is present in the `transaction_to_prof` mapping because there's a limit to how many rows you can return from snuba.

What you'll have to do here is to first query the span intervals, collect the list of transaction ids from there, then make a query for the profile id on the transactions table and filter on that list of transaction ids. (I'm not sure about the performance of this query though).

Reply:
@Zylphrex I don't think that's the case.

The `transaction_to_prof` contains the transaction IDs of (at most) 100 transactions that have a profile attached.

The `get_span_intervals` fetches the span intervals filtering on those transaction IDs (see `Condition(Column("transaction_id"), Op.IN, transaction_ids)`) so any transaction ID we get from the indexed spans would be in the mappings already. 

Does it make sense?